Geany updates

3eka opened the pull request:
This change contains.

- Update of Geany editor (version1.28)
- Addition of Geany editor plugins (version 1.28)

3eka committed:
Squashed?

Geany editor: bump to version 1.28

Geany editor: add plugins, version 1.28

Geany editor: add plugins, version 1.28/Fixes

Geany editor: move to components/editor; added REQUIRED_PACKAGES and fixes

Geany updates: removed geany-plugins; updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

Geany editor: add plugins, version 1.28

Geany editor: add plugins, version 1.28/Fixes

Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

3eka committed:
Geany editor: bump to version 1.28

3eka committed:
Geany editor: add plugins, version 1.28

Geany editor: add plugins, version 1.28/Fixes

Geany editor: move to components/editor; added REQUIRED_PACKAGES and fixes

Geany updates: removed geany-plugins; updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

3eka committed:
Geany editor: add plugins, version 1.28

3eka committed:
Geany editor: add plugins, version 1.28/Fixes

pyhalov commented on components/geany-plugins/Makefile:
```diff
@@ -0,0 +1,48 @@
+CFLAGS += -D_XOPEN_SOURCE=600
+
+# These 2 plugins fail to be compiled
+CONFIGURE_OPTIONS    += --disable-debugger --disable-scope
```
style: split it into 2 lines:
CONFIGURE_OPTIONS +=
CONFIGURE_OPTIONS +=

pyhalov commented on components/geany-plugins/Makefile:
```diff
@@ -0,0 +1,48 @@
+build: $(BUILD_32)
+
+install: $(INSTALL_32)
+
```
REQUIRED_PACKAGES?

pyhalov commented on components/geany/Makefile:
```diff
@@ -34,13 +34,10 @@ include ../../make-rules/ips.mk
 install: $(INSTALL_32)
 
 test: $(NO_TESTS)
-
```
REQUIRED_PACKAGES is missing

pyhalov commented:
Should this PR be split in two?

pyhalov commented:
Please, move components to appropriate category (editor).

3eka committed:
Geany editor: move to components/editor; added REQUIRED_PACKAGES and fixes

3eka commented:
Hi,
Guess I have fixed all (most) of problems...
I would like to keep this in same PR because I think that geany-plugins make no sense w/o geany itself.

Since this is my really first PR, I might made some mistakes....
Regards.

alarcher commented:
@3eka What Alexander meant is that PRs should be pushed in order of dependencies such that the build server processes them once at a time.
Given that geany-plugins may be compiled against the newer geany, it means that geany 1.28 should be updated, built and installed on the build server _before_ the PR containing geany-plugins is merged.
Does that explain

3eka commented:
@alarcher 
Thanks for info. If that is case I agree, **but** would need instructions how to do it now.
Also, for me it is unclear shall I do another fork or existing one is sufficient?
I will try to document everything so we can update outdated wiki page.

Regards.

alarcher commented:
@3eka The steps required are:
1) create another branch for geany-plugins,
2) remove the geany-plugins component from your geany branch,
3) if further modifications should be done, remove the geany commit from the geany-plugins branch.
4) open a 2nd pull request for geany-plugins

---

1) For safety let us first create the branch for geany-plugins.
Assuming that you are on your geany_updates branch:

$ git checkout -b geany_plugins

You can verify the list of commits using git log.

2) Go back to the 'geany_updates' branch:

$ git checkout geany_updates

Since you have 4 new commits listed, you need to change them:

git rebase -i HEAD~4

where '4' means that only the last 4 commits may be modified at your convenience.

Remove the 'geany-plugins' commit lines to keep only the update of geany to 1.28.
Save and exit: git will drop the commit corresponding to lines removed.
You can check with git log.

$ git push -f 3eka geany_updates

The '-f' (force) is required since we are rewriting the history of the remote branch.

3) Go to the 'geany_plugins' branch:

$ git checkout geany_plugins

Do the same procedure but this time drop the geany update commit and keep others.
Additionnally you can 'squash' them i.e merge several commits in one: to do so, replace 'pick' by 's' (keywords are described in the help when you rebase).

Verify that you have only one commit with the geany-plugins component.
Tip: in case you made a mistake and want to edit the message of the last commit, use:

$ git commit --amend

When you are happy:

$ git push 3eka geany_plugins

4) Open a second PR using the 'geany_plugins' branch as source.

3eka commented:
@alarcher 
Thanks again!

Looking to this guide, it looks to me that easier way is:
a) create geany_plugins branch from existing geany_updates
b) remove components/editor/geany-plugins from geany_updates branch and commit/push - this branch stays with geany editor only (so all applied modifications will stay)
c) remove components/editor/geany from geany_plugins branch and commit/push - this one will cover plugins only - i just need to create pull request from it (again, all modifications stay)

Would it be feasible?
Regards.

alarcher commented:
@3eka that's exactly what I described no?

3eka committed:
Geany updates: removed geany-plugins; updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

3eka committed:
Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES

alarcher commented:
@3eka You have to squash your commits anyway: by removing the files manually, you have done extra work.

3eka committed:
Merge branch 'geany_updates' of https://github.com/3eka/oi-userland into geany_updates

3eka commented:
squashed

pyhalov commented:
Don't see this. PR still has 8 commits.

3eka commented:
Not sure:

$ git checkout
Your branch is ahead of 'origin/geany_updates' by 2 commits.
  (use "git push" to publish your local commits)

$ git push 3eka geany_updates
Username for 'https://github.com': 3eka
Password for 'https://3eka@github.com':
Everything up-to-date

$ git checkout
Your branch is ahead of 'origin/geany_updates' by 2 commits.
  (use "git push" to publish your local commits)

How to fix this?
Thanks

3eka commented:
Was this proper fix:

$ git pull --rebase
From https://github.com/3eka/oi-userland
   865cdd0..c63d1af  geany_updates -> origin/geany_updates
Current branch geany_updates is up to date.

$ git checkout
Your branch is up-to-date with 'origin/geany_updates'.

If not, how to get out from this weird situation?
What would be best way (start from scratch)?

Regards.

alarcher commented:
@3eka There is no weird situation and starting from scratch is not a solution. You can follow the instructions I provided to you earlier.

3eka committed:
Deleted

3eka committed:
Sync
Merge branch 'geany_updates' of https://github.com/3eka/oi-userland into geany_updates

3eka commented:
OK,
got some time to try to fix this:

<pre>
$ git branch
  geany_plugins
* geany_updates
  oi/hipster
</pre>

<pre>
$ git checkout 
Your branch is ahead of 'origin/geany_updates' by 3 commits.
  (use "git push" to publish your local commits)
</pre>

So, what is next step (I am not sure anymore)? Sorry for being clumsy...
Regards.

alarcher commented:
git rebase -i HEAD~11

Change 'pick' to 's' for each line except the first and save+exit.

git push -f 3eka geany_updates

3eka commented:
@alarcher tried that:

<pre>
$ git rebase -i HEAD~11
pick db8fd92 curl: update to 7.51.0
squash 9a786bc Add sudosh2 1.0.5 (Predrag Zečević)
squash b184737 openjdk8: update to 1.8.112
squash 828773b nvidia: update to 340.98
squash f83241e OpenVPN bump to 2.3.12 and improve support for multi-instance setups (#2541)
squash 3622d90 ddu: package is not republished on Makefile change.
squash 48b1047 Add udpxy to oi-userland
squash 5944bda OpenVPN.xml : move the XML blocks according to schema
squash 62b8fdf Squashed?
squash fc77117 Deleted
squash 2f6e84e zlib Makefile : save the non-standard "configure" from arguments it does not support
squash 83d3827 Geany editor: bump to version 1.28
squash 589d8f3 Geany editor: add plugins, version 1.28
squash b726924 Geany editor: add plugins, version 1.28
squash 01444b2 Geany editor: add plugins, version 1.28/Fixes
squash fa018d4 Geany editor: move to components/editor; added REQUIRED_PACKAGES and fixes
squash 1afd23a Geany updates: removed geany-plugins; updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES
squash 865cdd0 Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES
</pre>

Saved it and got this back:

<pre>
error: could not apply 83d3827... Geany editor: bump to version 1.28
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
Could not apply 83d3827e0e071c2500d3b4d54c1c34e8a9405a9b... Geany editor: bump to version 1.28
</pre>

Regards.

alarcher commented:
@3eka Alright... your local branch does not look like the one you push (at least not on this PR interface).
Do we agree that "squashing" means "merging several of _your_ commits" ? Other commits should not be touched

git rebase --abort

I do not know what you have done with your branch but this should work ;)
Rebase against latest head of oi-userland

git pull --rebase origin oi/hipster

Count the number of your commits in the log (only the latests should be yours):

git log

Let us say you have N of _your_ commits to merge into one:

git rebase -i HEAD~N

Do the same thing as I explained earlier (put only 's' in front of _your_ commits not others)

git log

to check that only one commit remains.

3eka commented:
Arghh,
looks like I have messed it properly:

<pre>
$ git rebase --abort
</pre>

<pre>
$ git pull --rebase origin oi/hipster
From https://github.com/3eka/oi-userland
 * branch            oi/hipster -> FETCH_HEAD
First, rewinding head to replay your work on top of it...
Applying: Geany editor: bump to version 1.28
Applying: Geany editor: add plugins, version 1.28
Applying: Geany editor: add plugins, version 1.28/Fixes
Applying: Geany editor: move to components/editor; added REQUIRED_PACKAGES and fixes
Applying: Geany updates: removed geany-plugins; updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES
Applying: Geany updates: updated Makefile with COMONENT_DESCRIPTION and REQUIRED_PACKAGES
Applying: Geany editor: add plugins, version 1.28
Using index info to reconstruct a base tree...
A       components/geany/Makefile
A       components/geany/geany.license
A       components/geany/geany.p5m
A       components/geany/manifests/sample-manifest.p5m
Falling back to patching base and 3-way merge...
CONFLICT (rename/delete): components/libbonoboui/Makefile deleted in Geany editor: add plugins, version 1.28 and renamed in bd21a3f45ab0f59483f3b574dc4ca8a8b5ef6910. Version bd21a3f45ab0f59483f3b574dc4ca8a8b5ef6910 of components/libbonoboui/Makefile left in tree.
error: Failed to merge in the changes.
Patch failed at 0007 Geany editor: add plugins, version 1.28
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
</pre>

Huh?

alarcher commented:
@3eka it just says that the file has changed in the remote branch "origin oi/hipster". I do not know if you modified it by mistake in your tree.
In any case,

git status -uno

to see which files are in conflict.

Fix the conflict, git add files_which_are_in_conflict

then 

git rebase --continue

3eka commented:
@alarcher 
did this:

<pre>
$ git status -uno
rebase in progress; onto 65e00ec
You are currently rebasing branch 'geany_updates' on '65e00ec'.
  (fix conflicts and then run "git rebase --continue")
  (use "git rebase --skip" to skip this patch)
  (use "git rebase --abort" to check out the original branch)
Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add <file>..." to mark resolution)
        added by us:     components/libbonoboui/Makefile
no changes added to commit (use "git add" and/or "git commit -a")
</pre>

<pre>
$ git add components/libbonoboui/Makefile
</pre>

<pre>
$ git log | /usr/gnu/bin/grep -E "Author.*Predrag"
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zecevic - Unix Systems Administrator <predrag.zecevic@2e-systems.com>
Author: Predrag Zečević < <predrag.zecevic@2e-systems.com>
</pre>

<pre>
$ git rebase -i HEAD~6
It seems that there is already a rebase-apply directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
        git rebase (--continue | --abort | --skip)
If that is not the case, please
        rm -fr "/export/home/zeka/Hipster/oi-userland.20161118/.git/rebase-apply"
and run me again.  I am stopping in case you still have something
valuable there.
</pre>

This did not brought any progress:

<pre>
$ git rebase --continue
</pre>

Again, stuck.

alarcher commented:
@3eka why do you insist on not following the steps? ;)
I never wrote to "git rebase -i HEAD~6" since you are in the middle of rebase, I wrote that you should git rebase --continue once the conflict is solved.

Nevermind, reset --hard to the commit before your first commit and pull --rebase on

https://github.com/alarcher/oi-userland/tree/geany_updates

pyhalov commented:
Any news on this?

3eka commented:
Heh, @pyhalov 

I still did not had time to read complete git guide and have problem with hg popping up all time under my fat fingers... 

Right now I have booted my zone and 
<pre>
$ git branch
  geany_plugins
* geany_updates
  oi/hipster
</pre>
So, checking:
<pre>
$ git status -uno
On branch geany_updates
Your branch is up-to-date with 'origin/geany_updates'.
You are currently rebasing branch 'geany_updates' on '65e00ec'.
  (all conflicts fixed: run "git rebase --continue")

nothing to commit (use -u to show untracked files)
</pre>
If I understood properly, no conflicts are present, my branch is in sync with  origin oi/hipster one. Next:
<pre>
$ git rebase --continue
Applying: Geany editor: add plugins, version 1.28
No changes - did you forget to use 'git add'?
If there is nothing left to stage, chances are that something else
already introduced the same changes; you might want to skip this patch.

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
</pre>
Huh, all is up-to-date, so nothing to push either:
<pre>
$ git push 3eka geany_updates
Username for 'https://github.com': 3eka
Password for 'https://3eka@github.com':
Everything up-to-date
</pre>
If this was correct until now, what would be next step, or if not - what I am doing wrong?

Regards.

pyhalov commented:
Have you read this https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History ? It seems to be rather brief and helpful.

pyhalov commented:
Merged as #2638
